File: test/reference-output/importExample.fs
module ImportExample

open Thoth.Json.Net

type UsesImport =
    {
        Type: string
        Recruiter: Basic.Recruiter
    }

    static member Decoder: Decoder<UsesImport> =
        Decode.object (fun get ->
            {
                Type = get.Required.Field "type" (GotynoCoders.decodeLiteralString "UsesImport")
                Recruiter = get.Required.Field "recruiter" Basic.Recruiter.Decoder
            }
        )

    static member Encoder value =
        Encode.object
            [
                "type", Encode.string "UsesImport"
                "recruiter", Basic.Recruiter.Encoder value.Recruiter
            ]

type HoldsSomething<'t> =
    {
        HoldingField: 't
    }

    static member Decoder decodeT: Decoder<HoldsSomething<'t>> =
        Decode.object (fun get ->
            {
                HoldingField = get.Required.Field "holdingField" decodeT
            }
        )

    static member Encoder encodeT value =
        Encode.object
            [
                "holdingField", encodeT value.HoldingField
            ]

type StructureUsingImport =
    {
        Event: Basic.Event
    }

    static member Decoder: Decoder<StructureUsingImport> =
        Decode.object (fun get ->
            {
                Event = get.Required.Field "event" Basic.Event.Decoder
            }
        )

    static member Encoder value =
        Encode.object
            [
                "event", Basic.Event.Encoder value.Event
            ]

type UnionUsingImport =
    | CoolEvent of Basic.Event
    | Other of Basic.Person

    static member CoolEventDecoder: Decoder<UnionUsingImport> =
        Decode.object (fun get -> CoolEvent(get.Required.Field "data" Basic.Event.Decoder))

    static member OtherDecoder: Decoder<UnionUsingImport> =
        Decode.object (fun get -> Other(get.Required.Field "data" Basic.Person.Decoder))

    static member Decoder: Decoder<UnionUsingImport> =
        GotynoCoders.decodeWithTypeTag
            "type"
            [|
                "CoolEvent", UnionUsingImport.CoolEventDecoder
                "Other", UnionUsingImport.OtherDecoder
            |]

    static member Encoder =
        function
        | CoolEvent payload ->
            Encode.object [ "type", Encode.string "CoolEvent"
                            "data", Basic.Event.Encoder payload ]

        | Other payload ->
            Encode.object [ "type", Encode.string "Other"
                            "data", Basic.Person.Encoder payload ]

type AllConcrete =
    {
        Field: HoldsSomething<Basic.Either<Basic.Maybe<StructureUsingImport>, UnionUsingImport>>
    }

    static member Decoder: Decoder<AllConcrete> =
        Decode.object (fun get ->
            {
                Field = get.Required.Field "field" (HoldsSomething.Decoder (Basic.Either.Decoder (Basic.Maybe.Decoder StructureUsingImport.Decoder) UnionUsingImport.Decoder))
            }
        )

    static member Encoder value =
        Encode.object
            [
                "field", (HoldsSomething.Encoder (Basic.Either.Encoder (Basic.Maybe.Encoder StructureUsingImport.Encoder) UnionUsingImport.Encoder)) value.Field
            ]

File: test/reference-output/importExample.py
import json
import typing
from dataclasses import dataclass
from gotyno_validation import encoding, validation

from . import basic

@dataclass(frozen=True)
class UsesImport:
    type: typing.Literal['UsesImport']
    recruiter: basic.Recruiter

    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['UsesImport']:
        return validation.validate_interface(value, {'type': validation.validate_literal('UsesImport'), 'recruiter': basic.Recruiter.validate}, UsesImport)

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['UsesImport']:
        return validation.validate_from_string(string, UsesImport.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        return {'type': 'UsesImport', 'recruiter': basic.Recruiter.to_json(self.recruiter)}

    def encode(self) -> str:
        return json.dumps(self.to_json())

T = typing.TypeVar('T')
@dataclass(frozen=True)
class HoldsSomething(typing.Generic[T]):
    holdingField: T

    @staticmethod
    def validate(validate_T: validation.Validator[T]) -> validation.Validator['HoldsSomething[T]']:
        def validate_HoldsSomethingT(value: validation.Unknown) -> validation.ValidationResult['HoldsSomething[T]']:
            return validation.validate_interface(value, {'holdingField': validate_T}, HoldsSomething)
        return validate_HoldsSomethingT

    @staticmethod
    def decode(string: typing.Union[str, bytes], validate_T: validation.Validator[T]) -> validation.ValidationResult['HoldsSomething[T]']:
        return validation.validate_from_string(string, HoldsSomething.validate(validate_T))

    def to_json(self, T_to_json: encoding.ToJSON[T]) -> typing.Dict[str, typing.Any]:
        return {'holdingField': T_to_json(self.holdingField)}

    def encode(self, T_to_json: encoding.ToJSON[T]) -> str:
        return json.dumps(self.to_json(T_to_json))

@dataclass(frozen=True)
class StructureUsingImport:
    event: basic.Event

    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['StructureUsingImport']:
        return validation.validate_interface(value, {'event': basic.Event.validate}, StructureUsingImport)

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['StructureUsingImport']:
        return validation.validate_from_string(string, StructureUsingImport.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        return {'event': basic.Event.to_json(self.event)}

    def encode(self) -> str:
        return json.dumps(self.to_json())

class UnionUsingImport:
    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['UnionUsingImport']:
        return validation.validate_with_type_tags(value, 'type', {'CoolEvent': CoolEvent.validate, 'Other': Other.validate})

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['UnionUsingImport']:
        return validation.validate_from_string(string, UnionUsingImport.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        raise NotImplementedError('`to_json` is not implemented for base class `UnionUsingImport`')

    def encode(self) -> str:
        raise NotImplementedError('`encode` is not implemented for base class `UnionUsingImport`')

@dataclass(frozen=True)
class CoolEvent(UnionUsingImport):
    data: basic.Event

    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['CoolEvent']:
        return validation.validate_with_type_tag(value, 'type', 'CoolEvent', {'data': basic.Event.validate}, CoolEvent)

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['CoolEvent']:
        return validation.validate_from_string(string, CoolEvent.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        return {'type': 'CoolEvent', 'data': self.data.to_json()}

    def encode(self) -> str:
        return json.dumps(self.to_json())

@dataclass(frozen=True)
class Other(UnionUsingImport):
    data: basic.Person

    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['Other']:
        return validation.validate_with_type_tag(value, 'type', 'Other', {'data': basic.Person.validate}, Other)

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['Other']:
        return validation.validate_from_string(string, Other.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        return {'type': 'Other', 'data': self.data.to_json()}

    def encode(self) -> str:
        return json.dumps(self.to_json())

@dataclass(frozen=True)
class AllConcrete:
    field: HoldsSomething[basic.Either[basic.Maybe[StructureUsingImport], UnionUsingImport]]

    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['AllConcrete']:
        return validation.validate_interface(value, {'field': (HoldsSomething.validate (basic.Either.validate (basic.Maybe.validate StructureUsingImport.validate) UnionUsingImport.validate))}, AllConcrete)

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['AllConcrete']:
        return validation.validate_from_string(string, AllConcrete.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        return {'field': HoldsSomething.to_json(basic.Either.to_json(basic.Maybe.to_json(StructureUsingImport.to_json) UnionUsingImport.to_json))(self.field)}

    def encode(self) -> str:
        return json.dumps(self.to_json())